### PR TITLE
F/outside click component

### PR DESCRIPTION
**Pull request by tsargent**
This PR adds a higher-order component for determining "click outside."  It can be used for popups, modals, menus, etc.  

I used it here on the new SimpleSelect component, so that we can close the select options when we click outside.  This functionality was originally built right into SimpleSelect, and is now abstracted into this new wrapper.

-------------

#### onClickOutside
a required function that will be called when we click outside of the child component.

--------------

That's it!

To see it in action, you can just view the SimpleSelect component.  It's the same functionality, just abstracted out.

It also throws an informative error if that function doesn't exist in props.  Could be a good practice moving forward?

### CODE REVIEW

**Comment by jasonfellows** [comment]
Do we need to listen for events outside the component? It could be way more simple using z-index and a fixed element rendered behind the component. That way, it's all encapsulated.

```CSS
.modal-clear-bg {
  height: 100%;
  left: 0;
  position: fixed;
  top: 0;
  width: 100%;
}
```

Then set zIndex depending on component state:
```javascript
  baseZIndex() {
    return this.state.show ? 300 : 1;
  },
```

Then render the shown component in front of the invisible background with a click handler on it:
```javascript
  datePicker() {
    if (this.state.show) {
      return (
        <div>
          <div
            className='modal-clear-bg'
            onClick={this.hideDatePicker}
            style={{zIndex: this.baseZIndex()-2}}></div>
          <DatePicker
            zIndex={this.baseZIndex()-1}
            date={this.momentDate().toDate()}
            show={this.state.show}
            onChangeDate={this.onChangeDate} />
        </div>
      );
    }
  },
```

You can also set the z-index for the input depending on state. No listening for clicks on the entire document.

**Comment by tsargent** [comment]
@jasonfellows What about cases where you have a few selects on one page?  The user opens one, than opens another.  Opening the 2nd one should close the first.  This is handled with this PR.  It seems like you'd have to click twice with this fixed element solution, no?

**Comment by jasonfellows** [comment]
Only if clicking to close outside of the select component, which is much more rare.

File: src/js/components/forms/fields/simple-select.js [hunk 1]
```diff
@@ -181,11 +170,13 @@ export default React.createClass({
     let options = this.renderOptions();
 
     return (
-      <div className="simple-select relative">
-        <input type="hidden" name={this.props.name} value={this.state.value} disabled={this.props.disabled} />
-        {value}
-        {this.state.show_options ? options : false}
-      </div>
+      <ClickOutside onClickOutside={this.onClickOutside}>
+        <div className="simple-select relative">
```
Comment (tsargent) [inline src/js/components/forms/fields/simple-select.js]: wrap your JSX with the component here

**Comment by jasonfellows** [comment]
+1

**Comment by keithpops** [comment]
+1

**Comment by tsargent** [comment]
closing this and opening a new one to develop

**Pull request closed without merging**